connmgr/connmanager.go

```diff
@@ -0,0 +1,321 @@
+				}
+
+			case handleDisconnected:
+				if connReq, ok := cm.conns[msg.id]; ok {
+					connReq.stateMtx.Lock()
+					connReq.state = ConnDisconnected
+					connReq.stateMtx.Unlock()
+
+					if connReq.conn != nil {
+						connReq.conn.Close()
+					}
+					log.Debugf("Connection closed: %v", connReq.Addr)
+					delete(cm.conns, msg.id)
+
+					if cm.cfg.OnDisconnection != nil {
+						go cm.cfg.OnDisconnection(connReq)
+					}
+
+					cm.handleFailedConn(connReq, msg.retry)
+				}
+
+			case handleFailed:
+				connReq := msg.c
+				connReq.stateMtx.Lock()
+				connReq.state = ConnFailed
+				connReq.stateMtx.Unlock()
+				log.Debugf("Connection failed %v: %v", connReq.Addr, msg.err)
+
+				cm.handleFailedConn(connReq, true)
+			}
+
+		case <-cm.quit:
+			break out
+		}
+	}
+
+	cm.wg.Done()
+	log.Trace("Connection handler done")
+}
+
+// Connect assigns an id and dials a connection to the address of the
+// connection request.
+func (cm *ConnManager) Connect(c *ConnReq) {
+	if atomic.LoadInt32(&cm.stop) != 0 {
+		return
+	}
+	atomic.StoreUint64(&c.id, atomic.AddUint64(&cm.connReqCount, 1))
+	log.Debugf("Attempting connection: %v", c.Addr)
+	conn, err := cm.cfg.Dial("tcp", c.Addr)
+	if err != nil {
+		cm.requests <- handleFailed{c, err}
+	} else {
+		cm.requests <- handleConnected{c, conn}
+	}
+}
+
+// Disconnect disconnects the connection corresponding to the given connection
+// id. If permanent, the connection will be retried with an increasing backoff
+// duration.
+func (cm *ConnManager) Disconnect(id uint64) {
+	if atomic.LoadInt32(&cm.stop) != 0 {
+		return
+	}
+	cm.requests <- handleDisconnected{id, true}
+}
+
+// Remove removes the connection corresponding to the given connection
+// id from known connections.
+func (cm *ConnManager) Remove(id uint64) {
+	if atomic.LoadInt32(&cm.stop) != 0 {
+		return
+	}
+	cm.requests <- handleDisconnected{id, false}
+}
+
+// Start launches the connection manager and begins connecting to the network.
+func (cm *ConnManager) Start() {
+	log.Trace("Connection manager started")
+	cm.wg.Add(1)
+	go cm.connHandler()
+
+	if cm.cfg.GetNewAddress != nil {
+		for i := atomic.LoadUint64(&cm.connReqCount); i < uint64(cm.cfg.MaxOutbound); i++ {
+			go cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
+		}
+	}
+}
+
+// Wait blocks until the connection manager halts gracefully.
+func (cm *ConnManager) Wait() {
+	cm.wg.Wait()
+}
+
+// Stop gracefully shuts down the connection manager.
+func (cm *ConnManager) Stop() {
+	if atomic.AddInt32(&cm.stop, 1) != 1 {
+		log.Warnf("Connection manager already stopped")
+		return
+	}
+	close(cm.quit)
+	log.Trace("Connection manager stopped")
+}
+
+// New returns a new connection manager.
+// Use Start to start connecting to the network.
+func New(cfg *Config) (*ConnManager, error) {
+	if cfg.Dial == nil {
+		return nil, errors.New("Config: Dial cannot be nil")
+	}
+	// Default to sane values
+	if cfg.RetryDuration == 0 {
+		cfg.RetryDuration = time.Second
```

Comment:
Also, I'd prefer definitions for the defaults at the top of the file.  Same with the `.MaxOutbound` default below.